src/video/x11/SDL_x11pen.c

```diff
@@ -0,0 +1,430 @@
+            }
+        } else if (format_return == 16) {
+            Sint16 *numdata = (Sint16 *) output;
+            for (k = 0; k < to_copy; ++k) {
+                dest[k] = numdata[k];
+            }
+        } else {
+            SDL_memcpy(dest, output, sizeof(Sint32) * to_copy);
+        }
+        X11_XFree(output);
+        return to_copy;
+    }
+    return 0; /* type mismatch */
+}
+
+/* 32 bit vendor + device ID from evdev */
+static Uint32
+xinput2_pen_evdevid(_THIS, int deviceid)
+{
+    Sint32 ids[2];
+
+    pen_atoms_ensure_initialized(_this);
+
+    if (2 != xinput2_pen_get_int_property(_this, deviceid, pen_atoms.device_product_id, ids, 2)) {
+        return 0;
+    }
+    return ((ids[0] << 16) | (ids[1] & 0xffff));
+}
+
+
+/* Gets unique device ID (which will be shared for pen / eraser pairs) */
+static SDL_PenGUID
+xinput2_pen_get_guid(_THIS, int deviceid)
+{
+    SDL_PenGUID guid;
+    Uint32 evdevid = xinput2_pen_evdevid(_this, deviceid); /* also initialises pen_atoms  */
+
+    Uint32 guid_words_base[sizeof(guid.data) / sizeof(Uint32)];
+    Uint32 *guid_words = guid_words_base;
+    const size_t guid_words_total = sizeof(guid.data) / sizeof(Uint32);
+    Uint32 *guid_words_end = guid_words + guid_words_total;
+
+    /* pen_atoms was initialised by xinput2_pen_evdevid() earlier */
+    Atom vendor_guid_properties[] = { /* List of vendor-specific GUID sources */
+        /* Atom must not be None */
+        pen_atoms.wacom_serial_ids,
+        None /* terminator */
+    };
+    Atom *vendor_guid_it;
+
+    SDL_memset(guid_words, 0, sizeof(guid));
+
+    /* Always put the evdevid in the first four bytes */
+    *guid_words = evdevid;
+    guid_words += 1;
+
+    /* XInput2 does not offer a general-purpose GUID, so we try vendor-specific GUID information */
+    for (vendor_guid_it = vendor_guid_properties;
+         guid_words_end != guid_words /* space left? */
+             && *vendor_guid_it != None;
+         ++vendor_guid_it) {
+
+        const Atom property = *vendor_guid_it;
+        const int words_written = xinput2_pen_get_int_property(_this, deviceid, property, (Sint32 *) guid_words, guid_words_end - guid_words);
+
+        guid_words += words_written;
+    }
+
+    memcpy(guid.data, guid_words_base, sizeof(guid_words_base));
+    return guid;
+}
+
+/* Heuristically determines if device is an eraser */
+static SDL_bool
+xinput2_pen_is_eraser(_THIS, int deviceid, char* devicename)
+{
+    SDL_VideoData *data = (SDL_VideoData *) _this->driverdata;
+    char dev_name[PEN_ERASER_ID_MAXLEN];
+    int k;
+
+    pen_atoms_ensure_initialized(_this);
+
+    if (pen_atoms.wacom_tool_type != None) {
+        Atom type_return;
+        int format_return;
+        unsigned long num_items_return;
+        unsigned long bytes_after_return;
+        unsigned char *tooltype_name_info;
+
+        /* Try Wacom-specific method */
+        if (Success == X11_XIGetProperty(data->display, deviceid,
```

Comment:
The X11 provided `Success` seems inapplicable here. In `X11/X.h:350` `Success` is defined as `0`, however according to the Xlib documentation, `Status` (which X11_XIGetProperty returns) is:

> Some functions return Status, an integer error indication. If the function fails, it returns a zero. If the function returns a status of zero, it has not updated the return arguments. Because C does not provide multiple return values, many functions must return their results by writing into client-passed storage. By default, errors are handled either by a standard library function or by one that you provide. Functions that return pointers to strings return NULL pointers if the string does not exist. 

(https://x.org/releases/current/doc/libX11/libX11/libX11.html#Errors)

this is causing that the inner codeblock gets executed, whenever it isn't supposed to. This leads to `type_return` being `0` and thus a `nullptr` in form of `tooltype_name` will be passed to `SDL_strcasecmp` and since this function is not Nullpointer-aware, it causes a SEGFAULT whenever I launch your SDL pen demo with SDL_VIDEODRIVER=x11 (so that XWayland gets used).

Reply:
This is a weird one-- while I agree with your reading of the libX11 spec on `Status`, the `XIGetProperty` man page states:

> The function returns `Success` if it executes successfully.

(https://www.x.org/releases/X11R7.5/doc/man/man3/XIGetProperty.3.html)

This matches what I believe to be the a [mirror of the xorg implementation](https://github.com/freedesktop/xorg-libXi/blob/8b4ea802fa3ba769dc8335f0e72c6f0cb7c9c35e/src/XIProperties.c#L244) and also the behaviour of other XInput2 functions like [`XIQueryVersion()`](https://www.x.org/archive/X11R7.5/doc/man/man3/XIQueryVersion.3.html) that declare a return type of `Status` yet give a code example that checks against `Success`.

Looking around a bit more (e.g., at GDK), it seems that most `XIGetProperty()` clients do not trust the return value, so I think making the code more defensive is the way to go here (XWayland may have interpreted the specs differently, for all I know...)

Thank you for reporting!

Reply:
hmm, true. I've just ran Mutter on Xorg and there `XIGetProperty` succeeded (and wrote the correct data into the vars) but it still returned `0`. So yes, this seems like an implementation bug with XWayland (or XOrg/XFree86, depending on how you want to see the docs 🙃 ). I guess that is also the reason why GDK does not trust it :)

Reply:
I'd argue it's an ambiguity in the spec that calls for a spec revision, but when it comes to fixing other people's implementations, I'd go with the "more specific beats more general" interpretation of the specs-as-written... :-)